### PR TITLE
New Navigation Drawer and refactorings from NavigationPresenter to Architecture Components

### LINKED ISSUE #131: Migrate NavigationPresenter to Architecture Components

### PR DESCRIPTION
**Pull request by bengelhaupt**
Introduces a new navigation drawer seen in `MainActivity` and refactored `NavigationPresenter` to Architecture Components.

The navigation is now inflated from a menu file. This allows for a more robust selection of navigation elements beacuse now only the item's id has to be supplied instead of the position (which differed from flavor to flavor). Dynamically changing the menu is also much easier, because altering the `menu` object of the `NavigationView` and invalidating it, is sufficient to change the drawer items.
Because of these simplifications, the `NavigationDrawerFragment` and `NavigationDrawerAdapter` could be removed. This has the tradeoff, that icon resources have to be generated for the menu items.

The pictures below show the drawer for all flavors when logged in and when logged out (only openSAP).

&nbsp;&nbsp;
  

Fixes #131

### CODE REVIEW

File: app/src/main/java/de/xikolo/viewmodels/main/NavigationViewModel.kt [hunk 1]
```diff
@@ -0,0 +1,37 @@
...
+
+    var drawerSection: Int =
+        if (UserManager.isAuthorized) {
+            R.id.navigation_my_courses
+        } else {
+            R.id.navigation_all_courses
+        }
+
+    val user
+        get() = UserDao.Unmanaged.find(UserManager.userId)
```
Comment (tobiasrohloff) [inline app/src/main/java/de/xikolo/viewmodels/main/NavigationViewModel.kt]: You could provide a `findCurrent` method to simply return the logged in user.

File: app/src/main/res/drawable/bg_navigation_counter.xml [hunk 2]
```diff
@@ -1,5 +1,6 @@
 <?xml version="1.0" encoding="utf-8"?>
-<shape xmlns:android="http://schemas.android.com/apk/res/android" android:shape="rectangle">
-    <solid android:color="@color/apptheme_main"/>
-    <corners android:radius="2dp"/>
+<shape xmlns:android="http://schemas.android.com/apk/res/android"
+    android:shape="rectangle">
+    <solid android:color="@color/apptheme_main" />
+    <corners android:radius="5000dp" />
```
Comment (tobiasrohloff) [inline app/src/main/res/drawable/bg_navigation_counter.xml]: wtf?

**Comment by tobiasrohloff** [comment]
Neat stuff 🙏 thx!

**Comment by tobiasrohloff** [comment]
Probably the top item should state "Login" and not "Profil" if not logged in.

**Comment by tobiasrohloff** [comment]
And did you removed the mooc.house specific tinting? Check `TintUtil` and the mooc.house color resources.

File: app/src/main/java/de/xikolo/viewmodels/main/NavigationViewModel.kt (hunk 1 above)
Comment (bengelhaupt) [reply app/src/main/java/de/xikolo/viewmodels/main/NavigationViewModel.kt]: Introduced `UserDao.Unmanaged.current`.
The `LiveData`-Approach does not work as it does not observe the database query but a database object resulting in app crashes when this observed database object is deleted e.g. on logout.

<details>
<summary>Realm log</summary>
2019-07-05 21:22:26.013 32351-32351/de.xikolo.moochouse.debug E/AndroidRuntime: FATAL EXCEPTION: main
    Process: de.xikolo.moochouse.debug, PID: 32351
    java.lang.IllegalStateException: Accessing object of type  which has been invalidated or deleted
        at io.realm.internal.OsObject.nativeStartListening(Native Method)
        at io.realm.internal.OsObject.addListener(OsObject.java:118)
        at io.realm.ProxyState.addChangeListener(ProxyState.java:141)
        at io.realm.RealmObject.addChangeListener(RealmObject.java:499)
        at io.realm.RealmObject.addChangeListener(RealmObject.java:542)
        at io.realm.RealmObject.addChangeListener(RealmObject.java:443)
        at de.xikolo.extensions.RealmObjectLiveData.onActive(RealmExtensions.kt:14)
        at androidx.lifecycle.LiveData$ObserverWrapper.activeStateChanged(LiveData.java:418)
        at androidx.lifecycle.LiveData$LifecycleBoundObserver.onStateChanged(LiveData.java:376)
        at androidx.lifecycle.LifecycleRegistry$ObserverWithState.dispatchEvent(LifecycleRegistry.java:355)
        at androidx.lifecycle.LifecycleRegistry.forwardPass(LifecycleRegistry.java:293)
        at androidx.lifecycle.LifecycleRegistry.sync(LifecycleRegistry.java:333)
        at androidx.lifecycle.LifecycleRegistry.moveToState(LifecycleRegistry.java:138)
        at androidx.lifecycle.LifecycleRegistry.handleLifecycleEvent(LifecycleRegistry.java:124)
        at androidx.lifecycle.ReportFragment.dispatch(ReportFragment.java:122)
        at androidx.lifecycle.ReportFragment.onStart(ReportFragment.java:82)
        at android.app.Fragment.performStart(Fragment.java:2548)
        at android.app.FragmentManagerImpl.moveToState(FragmentManager.java:1334)
        at android.app.FragmentManagerImpl.moveFragmentToExpectedState(FragmentManager.java:1576)
        at android.app.FragmentManagerImpl.moveToState(FragmentManager.java:1637)
        at android.app.FragmentManagerImpl.dispatchMoveToState(FragmentManager.java:3046)
        at android.app.FragmentManagerImpl.dispatchStart(FragmentManager.java:3003)
        at android.app.FragmentController.dispatchStart(FragmentController.java:193)
        at android.app.Activity.performStart(Activity.java:7192)
        at android.app.ActivityThread.handleStartActivity(ActivityThread.java:3137)
        at android.app.servertransaction.TransactionExecutor.performLifecycleSequence(TransactionExecutor.java:180)
        at android.app.servertransaction.TransactionExecutor.cycleToPath(TransactionExecutor.java:165)
        at android.app.servertransaction.TransactionExecutor.executeLifecycleState(TransactionExecutor.java:142)
        at android.app.servertransaction.TransactionExecutor.execute(TransactionExecutor.java:70)
        at android.app.ActivityThread$H.handleMessage(ActivityThread.java:1945)
        at android.os.Handler.dispatchMessage(Handler.java:106)
        at android.os.Looper.loop(Looper.java:193)
        at android.app.ActivityThread.main(ActivityThread.java:7009)
        at java.lang.reflect.Method.invoke(Native Method)
        at com.android.internal.os.RuntimeInit$MethodAndArgsCaller.run(RuntimeInit.java:537)
        at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:858)
</details>

File: app/src/main/res/drawable/bg_navigation_counter.xml (hunk 2 above)
Comment (bengelhaupt) [reply app/src/main/res/drawable/bg_navigation_counter.xml]: Still better than a 9-patch :smile:

**Comment by bengelhaupt** [comment]
See comments above. `Profile` has been renamed to `Login`. The `mooc.house` specific resources for the drawer have been removed. The sticky-posting has been removed.

**Comment by bengelhaupt** [comment]
The reverting of posting the login and logout events sticky now makes the drawer not refresh upon one of these events for me (debug, any flavor). Please test on your device and eventually with a production release.

File: app/src/main/java/de/xikolo/viewmodels/main/NavigationViewModel.kt (hunk 1 above)
Comment (tobiasrohloff) [reply app/src/main/java/de/xikolo/viewmodels/main/NavigationViewModel.kt]: Ok, sad.

**Comment by tobiasrohloff** [comment]
When you look at the `BaseActivity` the register/unregister on the event bus is done onStart/onStop. Thus, the login/logout state change was never triggered with these events for the `MainActivity` when it is in background. They are either there for another case or legacy. You should try to update the state in the onStart method of the `MainActivity`.

**Comment by bengelhaupt** [comment]
Well, that was the catch. Thank you!

**Comment by tobiasrohloff** [comment]
The drawer updates itself fine now, but the main fragments not :( We need to investigate here. I was also planning to remove the events at all. What about simply restarting the MainActivity after a login/out. That would avoid the need for such state changes, or?

**Comment by tobiasrohloff** [comment]
Ok, I had another idea, but that would be kind of a new feature. If the user isn't logged in, after the SplashActivity another activity opens, but not the main one. Just a simple course list, which reuses our course list fragment. A click on enroll opens the login activity as usual.

**Comment by bengelhaupt** [comment]
> The drawer updates itself fine now, but the main fragments not

For me it works. The Fragents receive the events and update accordingly after login. The `My Courses`-List has a Heisenbug showing the `No Enrollments`-Error after updating. I have not been able to fix this to this point.

> What about simply restarting the MainActivity after a login/out. That would avoid the need for such state changes, or?

That's true, but we would still need some path of communication between login logic and the MainActivity. We could set a persistent flag in the preferences indicating that the MainActivity should `recreate()` itself.

> Ok, I had another idea, but that would be kind of a new feature. If the user isn't logged in, after the SplashActivity another activity opens, but not the main one. Just a simple course list, which reuses our course list fragment. A click on enroll opens the login activity as usual.

So you mean we have some sort of `UserLoggedInMainActivity` and `UserLoggedOutMainActivity`? I don't know if this would work. Once, the new activity would need to be a full-fledged MainActivity. What about Settings, Download, Global News (generally users without account)? Secondly, we would still somehow need to communicate which of the activities to use and set some sort of flag.

**Pull request merged**